File: BatchTestTool/java/java automated tester/Test/compile_passed/solutions/JobMarketplace.java
import java.util.ArrayList;
import java.util.List;

class Job {
    String jobTitle;
    String company;
    List<String> requirements;

    Job(String jobTitle, String company, List<String> requirements) {
        this.jobTitle = jobTitle;
        this.company = company;
        this.requirements = requirements;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Job other = (Job) obj;
        return jobTitle.equals(other.jobTitle) &&
                company.equals(other.company) &&
                requirements.equals(other.requirements);
    }
}

class Resume {
    String name;
    List<String> skills;
    String experience;

    Resume(String name, List<String> skills, String experience) {
        this.name = name;
        this.skills = skills;
        this.experience = experience;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Resume other = (Resume) obj;
        return name.equals(other.name) &&
                skills.equals(other.skills) &&
                experience.equals(other.experience);
    }
}

public class JobMarketplace {
    private List<Job> jobListings;
    private List<Resume> resumes;

    public JobMarketplace() {
        jobListings = new ArrayList<>();
        resumes = new ArrayList<>();
    }

    public void postJob(String jobTitle, String company, List<String> requirements) {
        jobListings.add(new Job(jobTitle, company, requirements));
    }

    public void removeJob(Job job) {
        jobListings.remove(job);
    }

    public void submitResume(String name, List<String> skills, String experience) {
        resumes.add(new Resume(name, skills, experience));
    }

    public void withdrawResume(Resume resume) {
        resumes.remove(resume);
    }

    public List<Job> searchJobs(String criteria) {
        List<Job> matchingJobs = new ArrayList<>();
        for (Job jobListing : jobListings) {
            boolean found = jobListing.jobTitle.contains(criteria);
            if (!found) {
                for (String requirement : jobListing.requirements) {
                    if (requirement.contains(criteria)) {
                        found = true;
                        break;
                    }
                }
            }

            if (found) {
                matchingJobs.add(jobListing);
            }
        }
        return matchingJobs;
    }

    public List<Resume> getJobApplicants(Job job) {
        List<Resume> applicants = new ArrayList<>();
        for (Resume resume : resumes) {
            if (matchesRequirements(resume, job.requirements)) {
                applicants.add(resume);
            }
        }
        return applicants;
    }

    public static boolean matchesRequirements(Resume resume, List<String> requirements) {
        for (String skill : resume.skills) {
            if (!requirements.contains(skill)) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        JobMarketplace marketplace = new JobMarketplace();

        // Example usage
        List<String> requirements = new ArrayList<>();
        requirements.add("Java");
        requirements.add("Spring");
        marketplace.postJob("Software Engineer", "Tech Corp", requirements);

        List<String> skills = new ArrayList<>();
        skills.add("Java");
        skills.add("Spring");
        marketplace.submitResume("John Doe", skills, "5 years");

        List<Job> matchingJobs = marketplace.searchJobs("Software");
        for (Job job : matchingJobs) {
            System.out.println("Job Title: " + job.jobTitle);
        }

        List<Resume> applicants = marketplace.getJobApplicants(matchingJobs.get(0));
        for (Resume applicant : applicants) {
            System.out.println("Applicant: " + applicant.name);
        }
    }
}

File: ClassEval_T/cpp/solution/JobMarketplace.cpp
#include <string>
#include <vector>
#include <algorithm>

struct Job {
    std::string job_title;
    std::string company;
    std::vector<std::string> requirements;
    bool operator==(const Job& other) const {
        return job_title == other.job_title &&
            company == other.company &&
            requirements == other.requirements;
    }
};

struct Resume {
    std::string name;
    std::vector<std::string> skills;
    std::string experience;
    bool operator==(const Resume& other) const {
        return name == other.name &&
            skills == other.skills &&
            experience == other.experience;
    }
};

class JobMarketplace {
public:
    JobMarketplace();

    void post_job(const std::string& job_title, const std::string& company, const std::vector<std::string>& requirements);
    void remove_job(const Job& job);
    void submit_resume(const std::string& name, const std::vector<std::string>& skills, const std::string& experience);
    void withdraw_resume(const Resume& resume);
    std::vector<Job> search_jobs(const std::string& criteria) const;
    std::vector<Resume> get_job_applicants(const Job& job) const;

    static bool matches_requirements(const Resume& resume, const std::vector<std::string>& requirements);

    std::vector<Job> job_listings;
    std::vector<Resume> resumes;
};


JobMarketplace::JobMarketplace() {}

void JobMarketplace::post_job(const std::string& job_title, const std::string& company, const std::vector<std::string>& requirements) {
    job_listings.push_back({ job_title, company, requirements });
}

void JobMarketplace::remove_job(const Job& job) {
    auto it = std::remove(job_listings.begin(), job_listings.end(), job);
    job_listings.erase(it, job_listings.end());
}

void JobMarketplace::submit_resume(const std::string& name, const std::vector<std::string>& skills, const std::string& experience) {
    resumes.push_back({ name, skills, experience });
}

void JobMarketplace::withdraw_resume(const Resume& resume) {
    auto it = std::remove(resumes.begin(), resumes.end(), resume);
    resumes.erase(it, resumes.end());
}

std::vector<Job> JobMarketplace::search_jobs(const std::string& criteria) const {
    std::vector<Job> matching_jobs;
    for (const auto& job_listing : job_listings) {
        bool found = (job_listing.job_title.find(criteria) != std::string::npos);
        if (!found) {
            for (const auto& requirement : job_listing.requirements) {
                if (requirement.find(criteria) != std::string::npos) {
                    found = true;
                    break;
                }
            }
        }

        if (found) {
            matching_jobs.push_back(job_listing);
        }
    }
    return matching_jobs;
}

std::vector<Resume> JobMarketplace::get_job_applicants(const Job& job) const {
    std::vector<Resume> applicants;
    for (const auto& resume : resumes) {
        if (matches_requirements(resume, job.requirements)) {
            applicants.push_back(resume);
        }
    }
    return applicants;
}

bool JobMarketplace::matches_requirements(const Resume& resume, const std::vector<std::string>& requirements) {
    for (const auto& skill : resume.skills) {
        if (std::find(requirements.begin(), requirements.end(), skill) == requirements.end()) {
            return false;
        }
    }
    return true;
}
